doc/Tools/MultidimensionalSparseTuningProblem.md

```diff
@@ -0,0 +1,47 @@
+# MultidimensionalSparseTuningProblem
+
+```
+What water is there for us to clean ourselves? What festivals of atonement, what sacred games shall we have to invent?
+```
+
+\- Nietsche and also Chidi Anagonye, foreseeing this godawful hack
+
+Okay, so MultidimensionalSparseTuningProblem is a _monster_. What I'd like to do, is talk it through first as a concept, and then as code.
+
+## Motivation/Why on earth...
+
+One critical part of any autotuning system is in how you describe constraints. Maybe the team size times the vector length on a TeamPolicy must be less than 1024. Maybe the product of all tile dimensions have to be less than 4096. Kokkos solves this by... not having a system for that at all. Oops. Snark aside, the tools we were looking to support did not all support constraint systems, and those that did had wildly different ideas of how to express constraints, and exposing a system to users that was portable across tools looked really difficult. So I made the call not to support constraints in the callback layer. This is one of the most questionable decisions in the Tools design, if you're looking for something flashy to make an impact with, adding a constraint system that doesn't exclude any of our existing tools (Apollo/CConfigSpace/APEX) would be neat. That said, MultidimensionalSparseTuningProblem is a hack to support constraints _above_ the level of the callback layer.
+
+## Modeled constraints
+
+So the overwhelming majority of constraints for tuning on our accelerators take the form "the product of some N dimensional set of choices must be less than X." But in general, MultidimensionalSparseTuningProblem can model any constraint where the first N choices you have made yields some finite set of valid values for the next choice, and all paths through the system involve the same number of choices. Different choices in earlier decisions might lead to different options in later ones, for example in the MDRange case (dimensions must multiply to less than 4096), a choice of 1 for the first tile size will yield many options for the second one, but a choice of 4096 will yield only 1 as an option. What's hard about this, though, is that you can't just say "my outer value is 256, my inner value is 16," because without spelling out constraints we don't knwo that 16 is a valid second option for 256.
```

Comment:
```suggestion
So the overwhelming majority of constraints for tuning on our accelerators take the form "the product of some N-dimensional set of choices must be less than X." But in general, MultidimensionalSparseTuningProblem can model any constraint where the first N choices you have made yield some finite set of valid values for the next choice, and all paths through the system involve the same number of choices. Different choices in earlier decisions might lead to different options in later ones, for example in the MDRange case (dimensions must multiply to less than 4096), a choice of 1 for the first tile size will yield many options for the second one, but a choice of 4096 will yield only 1 as an option. What's hard about this, though, is that you can't just say "my outer value is 256, my inner value is 16," because without spelling out constraints we don't know that 16 is a valid second option for 256.
```